ibverbs/src/lib.rs

```diff
@@ -533,6 +553,97 @@ impl<'ctx> CompletionQueue<'ctx> {
+    ///
+    /// Unlike `poll`, this method blocks until at least one work completion is available or the
+    /// optional timeout expires. It is designed to wait efficiently for completions when polling
+    /// alone is insufficient, such as in low-traffic scenarios.
+    ///
+    /// The returned slice reflects completed work requests (e.g., sends, receives) from the
+    /// associated Work Queue. Not all fields in `ibv_wc` are valid unless the status is
+    /// `IBV_WC_SUCCESS`.
+    ///
+    /// # Errors
+    /// - `TimedOut`: If the timeout expires before any completions are available.
+    /// - System errors: From underlying calls like `req_notify_cq`, `poll`, or `ibv_get_cq_event`.
+    pub fn wait<'c>(
+        &self,
+        completions: &'c mut [ffi::ibv_wc],
+        timeout: Option<Duration>,
+    ) -> io::Result<&'c mut [ffi::ibv_wc]> {
+        let c = completions as *mut [ffi::ibv_wc];
+
+        //
+        loop {
+            let polled_completions = self.poll(unsafe { &mut *c })?;
+            if !polled_completions.is_empty() {
+                return Ok(polled_completions);
+            }
+
+            // SAFETY: dereferencing completion queue, which is guaranteed to not have been destroyed yet.
+            let ctx = unsafe { *self.cq }.context;
+            let errno = unsafe {
+                let ops = &mut { &mut *ctx }.ops;
+                ops.req_notify_cq.as_mut().unwrap()(self.cq, 0)
+            };
+            if errno != 0 {
+                return Err(io::Error::from_raw_os_error(errno));
+            }
+
+            // We poll again to avoid a race when Work Completions arrive between the first `poll()` and `req_notify_cq()`.
+            let polled_completions = self.poll(unsafe { &mut *c })?;
+            if !polled_completions.is_empty() {
+                return Ok(polled_completions);
+            }
+
+            // ibv_get_cq_event supports blocking operations, but the fd of cq_context was put into non blocking mode to support timeouts.
+            let pollfd = nix::poll::PollFd::new(
+                // SAFETY: dereferencing completion queue context, which is guaranteed to not have been destroyed yet.
+                unsafe { BorrowedFd::borrow_raw({ *self.cc }.fd) },
+                nix::poll::PollFlags::POLLIN,
+            );
+            let ret = nix::poll::poll(
+                &mut [pollfd],
+                timeout
+                    .map(nix::poll::PollTimeout::try_from)
+                    .transpose()
+                    .map_err(|_| io::Error::other("failed to convert timeout to PollTimeout"))?,
+            )?;
+            match ret {
+                0 => {
+                    return Err(io::Error::new(
+                        io::ErrorKind::TimedOut,
+                        "Timed out during completion queue wait",
+                    ))
+                }
+                1 => {}
+                _ => unreachable!(),
```

Comment:
```suggestion
                _ => unreachable!("we passed 1 fd to poll, but it returned {ret}"),
```
so that we get something useful out if this ever _does_ trigger